Add watch option for bluechi status

ArtiomDivak opened the pull request:
This PR will add watch option for 'bluechi status'. I added monitor for the units that are requested to be watch. And every time the status of the service changed the screen is cleared and every the new status is printed .

Related: https://github.com/eclipse-bluechi/bluechi/issues/706

ArtiomDivak committed:
Add watch option for bluechi status

This PR will add watch option for 'bluechi status'.
I added monitor for the units that are requested to be watch.
And every time the status of the service changed the screen is cleared
and every the new status is printed .

Related: https://github.com/eclipse-bluechi/bluechi/issues/706
Signed-off-by: Artiom Divak <adivak@redhat.com>

engelmi commented on src/client/method-status.c:
```diff
@@ -282,9 +303,106 @@ static int method_status_unit_on(Client *client, char *node_name, char **units,
+                        if (r < 0) {
+                                fprintf(stderr, "Failed closing subscription unit array: %s\n", strerror(-r));
+                                return r;
+                        }
```
The `bluechictl status <node> <unit>` is only for one unit, so no need to subscribe to a list of units. The code becomes a lot smaller then as well:
```suggestion
                        _cleanup_sd_bus_message_ sd_bus_message *m = NULL;
                        r = sd_bus_message_new_method_call(
                                        client->api_bus,
                                        &m,
                                        BC_INTERFACE_BASE_NAME,
                                        monitor_path,
                                        MONITOR_INTERFACE,
                                        "Subscribe");
                        if (r < 0) {
                                fprintf(stderr, "Failed creating subscription call: %s\n", strerror(-r));
                                return r;
                        }

                        sd_bus_message_append(m, "ss", node_name, unit_name);
```

ArtiomDivak replied:
OK But is the for loop above to get multiple unit statuses?

engelmi replied:
As we discussed, there is currently a bug in `bluechictl status <node> <unit1> <unit2>...` - I created issue https://github.com/eclipse-bluechi/bluechi/issues/869. 
@ArtiomDivak Could you fix this (and maybe include the other memory leak) first and then continue on this PR?

engelmi commented on src/client/method-status.c:
```diff
@@ -282,9 +303,106 @@ static int method_status_unit_on(Client *client, char *node_name, char **units,
+                                return r;
+                        }
+
+                        r = sd_bus_call(client->api_bus, m, BC_DEFAULT_DBUS_TIMEOUT, &error, &reply);
```
Reusing `reply` here will cause a memory leak since it hasn't been freed after its first usage a few lines above. Therefore, I'd just declare another sd_bus_message variable:
```suggestion
                        _cleanup_sd_bus_message_ sd_bus_message *subscription_reply = NULL;
                        r = sd_bus_call(client->api_bus, m, BC_DEFAULT_DBUS_TIMEOUT, &error, &subscription_reply);
```

engelmi commented on src/client/method-status.c:
```diff
@@ -266,7 +273,21 @@ static int get_status_unit_on(Client *client, char *node_name, char *unit_name,
+static int on_unit_status_changed(UNUSED sd_bus_message *m, void *userdata, UNUSED sd_bus_error *error) {
+        units_status *user_data = (units_status *) userdata;
+        printf("\033[2J");
+        printf("\033[%d;%dH", 0, 0);
```
I think it would be great to move these two lines to a dedicated function `clear_screen()` in `method-status.c` and use it here as well as in other places - this way its more descriptive what happens (the escape codes are a bit tricky to read).

engelmi commented on src/client/method-status.c:
```diff
@@ -266,7 +273,21 @@ static int get_status_unit_on(Client *client, char *node_name, char *unit_name,
+        printf("\033[2J");
+        printf("\033[%d;%dH", 0, 0);
+        print_info_header(*user_data->max_name_len);
+        get_status_unit_on(
```
Just as an FYI: 
Its a bit unfortunate, but I guess we can't avoid having a dedicated call to fetch the latest state since the signal does not contain information about the `enabled` or `freeze` state.

engelmi commented on src/client/method-status.c:
```diff
@@ -266,7 +273,21 @@ static int get_status_unit_on(Client *client, char *node_name, char *unit_name,
+        printf("\033[%d;%dH", 0, 0);
+        print_info_header(*user_data->max_name_len);
+        get_status_unit_on(
+                        user_data->client, user_data->node_name, user_data->unit_name, *user_data->max_name_len);
```
You don't need the `struct unit_status`. The node and unit name are in the signal message m anyway and since its only one unit the max len is the length of that string. Passing only the `Client* client` in the userdata is sufficient, for example:

```suggestion
        Client *client = (Client *) userdata;
        clear_screen();
               
        /*
        read information from signal message m
        - node_name
        - unit_name 
        the other information (state and substate can be omitted since we fetch it anyway again)
        */ 
        
        // there is only one unit, so max_name_len is the length of the unit name
        size_t max_name_len = strlen(unit_name);        
        print_info_header(max_name_len);
        get_status_unit_on( node_name, unit_name, max_name_len);
```

__Note: __ 
By removing the `unit_status` struct the memory management becomes easier. At the moment, there is a memory leak due to the `malloc0(sizeof(units_status))` not being freed.

engelmi commented on src/client/method-status.c:
```diff
@@ -21,6 +21,13 @@ struct bus_properties_map {
         size_t offset;
```
Not part of this PR, but would be great to fix it here:

In [src/client/client.c#115](https://github.com/eclipse-bluechi/bluechi/blob/main/src/client/client.c#L115), a call to `assemble_object_path_string` is being made and the result is assigned to `client->object_path`. This causes a memory leak when called more than once since `client->object_path` doesn't get freed before. So adding the following line there (in a separate commit) would fix that issue:
```C
// src/client/client.c#114-115 (client_create_message_new_method_call)
free_and_null(client->object_path)
r = assemble_object_path_string(NODE_OBJECT_PATH_PREFIX, node_name, &client->object_path);
```

engelmi reviewed (commented):
If you do any changes in BlueChi's C code, using valgrind is a great way to make sure that no memory leaks were introduced. In case of this PR, running valgrind on top of bluechictl is sufficient, e.g.
```bash
sudo valgrind --leak-check=full ./builddir/bin/usr/local/bin/bluechictl status laptop cow.service -w
```
Since valgrind performs a runtime analysis its important to use the new compiled binary (bluechictl) so that the changes are executed.

ArtiomDivak committed:
A call to assemble_object_path_string is being made and the result is assigned to client->object_path.
This causes a memory leak when called more than once since client->object_path doesn't get freed before.
So adding the following line should fix it.

Signed-off-by: Artiom Divak <adivak@redhat.com>

mkemel commented on src/client/client.c:
```diff
@@ -116,7 +116,8 @@ int client_create_message_new_method_call(
         _cleanup_sd_bus_message_ sd_bus_message *outgoing_message = NULL;
 
-        r = assemble_object_path_string(NODE_OBJECT_PATH_PREFIX, node_name, &client->object_path);
+        free_and_null(client->object_path) r = assemble_object_path_string(
```
Missing a newline here?

mwperina replied:
Right, newline and semicolon seem to be missing

mwperina commented on src/client/method-status.c:
```diff
@@ -11,6 +11,8 @@
 #include "libbluechi/common/time-util.h"
 
+// #include "libbluechi/service/shutdown.h"
+
```
Please remove commented lines

mwperina commented on src/client/method-status.c:
```diff
@@ -25,6 +27,25 @@ struct bus_properties_map {
+
+void clear_screen() {
+        printf("\033[2J");
+        printf("\033[%d;%dH", 0, 0);
```
Are we sure that this works on all terminals? Or at least those standard one?

engelmi replied:
These are [ANSI escape codes](https://en.wikipedia.org/wiki/ANSI_escape_code). They should be supported pretty much everywhere.

mwperina replied:
OK

mwperina commented on tests/tests/tier0/bluechi-agent-set-loglevel/main.fmf:
Why do you remove this file?

engelmi replied:
It seems like `bluechi-agent-set-loglevel` was used as a template. 
The `.fmf` file has been deleted and the `.py` was moved to the new test directory.

Please restore
- `tests/tests/tier0/bluechi-agent-set-loglevel/main.fmf`
- `tests/tests/tier0/bluechi-agent-set-loglevel/test_bluechi_agent_set_loglevel.py`

mwperina commented on tests/tests/tier0/bluechi-client-watch-option/python/bluechictl_status_watch.py:
```diff
@@ -0,0 +1,56 @@
+
+    def setUp(self) -> None:
+        service_status_active = re.compile(
+            r"^" + re.escape(simple_service) + ".* active"
```
What character do we need to escape in "simple.service"?

mwperina commented on tests/tests/tier0/bluechi-client-watch-option/test_bluechi_client_watch_option.py:
Did you really wanted to move that file?

engelmi commented on src/client/method-status.c:
```diff
@@ -293,9 +344,83 @@ static int method_status_unit_on(Client *client, char *node_name, char **units,
+                                        strerror(-r));
+                                return r;
+                        }
+                        //_cleanup_method_status_change_
```
Please remove this commented line.

engelmi commented on tests/tests/tier0/bluechictl-status-watch-service/python/bluechictl_status_watch.py:
```diff
@@ -0,0 +1,52 @@
+        except DBusError:
+            pass
+        self.bgrunner.stop()
+        assert self.evaluator.processing_finished()
```
The comment does not apply to this test so it should be removed. 
The same applies to the try-except. It is not needed - skipping any thrown error is actually bad here since it covers up a bug.  
```suggestion
        node = Node(node_foo_name)
        node.stop_unit(simple_service, "replace")
        node.start_unit(simple_service, "replace")
        node.stop_unit(simple_service, "replace")
        
        self.bgrunner.stop()
        assert self.evaluator.processing_finished()
```

engelmi commented on tests/tests/tier0/bluechictl-status-watch-service/test_bluechictl_status_watch_service.py:
```diff
@@ -3,48 +3,43 @@
-    assert "INFO" in output
+    node_foo.install_systemd_service(service)
+    print(ctrl.bluechictl.daemon_reload_node(NODE_FOO)[1])
+    print(ctrl.bluechictl.start_unit(NODE_FOO, service.name)[1])
```
After restoring the test files (see [this comment](https://github.com/eclipse-bluechi/bluechi/pull/865/files#r1570502722)), please remove these wrapping prints. 
```suggestion
    ctrl.bluechictl.daemon_reload_node(NODE_FOO)[1]
    ctrl.bluechictl.start_unit(NODE_FOO, service.name)[1]
```

engelmi commented on tests/tests/tier0/bluechictl-status-watch-service/test_bluechictl_status_watch_service.py:
```diff
@@ -3,48 +3,43 @@
 #
 # SPDX-License-Identifier: LGPL-2.1-or-later
 
+import logging
```
No logging happening here (the `logging.getLogger` prevents a lint error). So please remove this import and line getting the logger.

engelmi commented on src/client/method-status.c:
```diff
@@ -25,6 +27,25 @@ struct bus_properties_map {
 
+typedef struct method_status_change method_status_change;
+
+struct method_status_change {
```
Our coding convention for structs is PascalCase - please change to `MethodStatusChange`.

engelmi commented on src/client/method-status.c:
```diff
@@ -277,10 +298,40 @@ static int get_status_unit_on(Client *client, char *node_name, char *unit_name,
         unsigned i = 0;
 
         size_t max_name_len = get_max_name_len(units, units_count);
+        method_status_change *s = NULL;
```
The simple `free` call in `free_method_status_change` is fine since all fields in `struct method_status_change` that would required free are owned elsewhere (e.g. `*client` is owned by "main") and are freed there. Therefore, ShutdownHook is an overkill for simply freeing `method_status_change *s` here. The following is sufficient:
```C
_cleanup_free_ method_status_change *s = NULL;
```

Please: 
- [x] use `_cleanup_free_` here
- [x] remove the usage of `ShutdownHook` (revert changes in client.[c|h], include, etc.)

engelmi commented on src/client/method-status.c:
```diff
@@ -293,9 +344,83 @@ static int method_status_unit_on(Client *client, char *node_name, char **units,
+                        s->max_len = max_name_len;
+                        s->node_name = node_name;
+                        s->units = units;
+                        s->units_count = units_count;
```
This `malloc0` causes a memory leak for multiple units. Subsequent iterations of the for-loop will cause `s` pointing to the newly allocated memory and the previously allocated memory is lost. This command should show you the leak:
```bash
valgrind --leak-check=full bluechictl status laptop simple.service other.service -w
```

Please move this block (from `s = malloc0(..)` to `s->units_count...`) up right below the declaration of `method_status_change *s = NULL`. This way, its only allocated once. No need for an `if(do_watch)` check... its just a small allocation and assigning some pointers.

mwperina commented on src/client/method-status.c:
```diff
@@ -277,10 +296,49 @@ static int get_status_unit_on(Client *client, char *node_name, char *unit_name,
+        // s->max_len = max_name_len;
+        // s->node_name = node_name;
+        // s->units = units;
+        // s->units_count = units_count;
```
Please remove commented code

engelmi commented on src/client/method-status.c:
```diff
@@ -25,6 +25,25 @@ struct bus_properties_map {
+
+void free_method_status_change(MethodStatusChange *s) {
+        free(s);
+}
```
This function can be removed now.

engelmi commented on src/client/method-status.c:
```diff
@@ -277,10 +296,49 @@ static int get_status_unit_on(Client *client, char *node_name, char *unit_name,
+        _cleanup_free_ MethodStatusChange *s = malloc0(sizeof(MethodStatusChange));
+        if (s == NULL) {
+                fprintf(stderr, "Failed to malloc memory for MethodStatusChange");
+                return 1;
```
```suggestion
                return ENOMEM;
```

mwperina reviewed (approved):
LGTM

engelmi reviewed (approved):
LGTM
